### ISSUE TITLE
Improve serialization

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#186]
There are a bunch of tradeoffs regarding serialization/deserialization.

The current strategy is:
- to Encode, we maintain one Encode buffer per client. When we want to buffer a message to a single individual client, we start by encoding it as Bytes. When we are reading to send messages (every `send_interval`), we gather all the messages that we need to send in each channel. We provide their size to the priority manager that will only try to add the messages from highest priority to lowest priority until the bandwidth is full. (we know each message size because they've been encoded as bytes already). Then we pack those individual messages into Packets (with continue bits for channel changes: "are there more messages for this channel" and "is there another channel"?) of size 1200 bytes max
- the reason why we want to limit the size of packets to 1200 bytes max (when we can; some individual messages might be of size bigger than 1200 bytes, for example if the replication_group is big) is because fragmented packets can introduce higher latency/packet-loss. If a single fragment is lost, the entire packet is lost!

We could switch to the next version of bitcode.
- much faster, data is much more compressible (for example a Vec {a, b, c} will be encoded as aaaaabbbbbccccc instead of abcabcabc... ), no need for serialization hints, latest version.
- how does it work? this time there is no bit-padding, instead each 'section' (for example aaaaa or bbbbb will be byte-padded). There is no easy way to check the number of bits used
**- potential strategy A:**
  - we order the messages by order of priorities (highest priority first), we then try to encode messages[0..k], where we multiply k by 2, until we reach the rate limiter. (Then we can do binary search to find exactly how to stay under the rate limiter?) (do we run compression before the rate limiter?)
  - now we have identified the messages we want to send. We can:
     - 1) encode messages[0..n] directly, which gives us a Bytes. If it's too big, we split it into a FragmentedPacket. The problem is that now we don't have control over creating fragmented packets or not. But is fragmentation really a problem? TCP games work fine even with it.
     - 2) do another binary search strategy to serialize data up to packets of size MTU. Individual packets that are bigger than MTU might have to be handled separately. This time we can avoid fragmentation but there's some wasted CPU where we encode stuff multiple times. Do we run compression as well?
     - Maybe start with 1) first, because packets might very rarely exceed MTU size? Especially with bitcode's efficient compression.
**- potential strategy B:**
  - we encode every message first individually and get the resulting size. The size of concatenating all messages individually will be much bigger than the final size of encoding messages[0..n] directly, because with multiple messages we get better padding, better compression, etc. We can then do everything similarly to what we do now.

**Comment by msvbg** [comment]
What do you mean by "rate limiter" here? Is it the MTU limit?

It seems like the primary problem is that you can't predict the size of encoded messages in the new version of bitcode. I would probably opt for a simple strategy for the time being, either the fragmenting packets or the naively encoding each message individually, with an opportunity to optimize further later on.

It generally feels like a good idea to be on the newer version of bitcode and not vendor it.

**Comment by cBournhonesque** [comment]
By Rate Limiter, I mean that there's an optional to add a maximum bandwidth usage: [book](https://cbournhonesque.github.io/lightyear/book/concepts/advanced_replication/bandwidth_management.html)

In general
- each message gets serialized immediately upon buffering: [code](https://github.com/cBournhonesque/lightyear/blob/main/lightyear/src/packet/message_manager.rs#L121)

(I serialize immediately because:
- it might be necessary to clone the message, for example if we need to retry sending in reliable channels. By serializing into `Bytes`, the clone is basically free. For example [here](https://github.com/cBournhonesque/lightyear/blob/23551c94c5d4783a3b8db41de54b98a62fa4db6c/lightyear/src/channel/senders/reliable.rs#L204-L204)
- it gives me access to the encoded size of the message, which can be useful. For example it lets me know right away if the message will need to be fragmented. However it's not really optimal to serialize each message individually like this (both for CPU time and for the size of the final output) as this [benchmark](https://github.com/cBournhonesque/lightyear/pull/188) shows. Maybe there's a way to avoid this?)

Then if rate limiting is enabled:
- I sort all messages from highest priority to lowest: [code](https://github.com/cBournhonesque/lightyear/blob/main/lightyear/src/packet/priority_manager.rs#L157)
- then I provide the message size to the rate limiter, up until the rate limiter is full (max bandwidth reached)

**Comment by msvbg** [comment]
@cBournhonesque Thanks for the context! I'm trying to become more familiar with the code and hope to contribute in the future :)

**Comment by n1ght-hunter** [comment]
i wonder if it would be worth looking at using https://github.com/rkyv/rkyv instead of bitcode

**Comment by caniko** [comment]
Rykv wins in zero-copy @cBournhonesque, also how do you plan to integrate this with webtransport? I don't think `wtransport` supports either `rykv` or `bitcode`

**Comment by cBournhonesque** [comment]
Yes I wanted to try rkyv, but i think it would be a pretty big change and serialization performance is not really my focus right now. I think `wtransport` is orthogonal from the serialization strategy, it just transports raw bytes. 
Actually past versions of lightyear were using bitcode

**Comment by cBournhonesque** [comment]
Yes ideally there would be a clean separation/abstraction between serialization and the rest of the code so any serialization method could be swapped in/out

**Comment by caniko** [comment]
#putitinthebacklog

**Comment by cBournhonesque** [comment]
Plan:
- add a compression step in lightyear_transport that happens before the final packets are created.
- the compression step uses a packer algorithm to make sure that the final result fits inside the MTU

**Comment by cBournhonesque** [comment]
Added compression where the packet-builder is compression aware and will try to put as many compressed packets as possible in the packet.
